### PR TITLE
Prophetnet optimization

### PR DESCRIPTION
**Pull request by guillaume-be**
# What does this PR do?

This PR proposes an optimization for the ProphetNet model. The current implementation calculates an attention bias mask by looping through the position to unmask. It performs a high number of assignments (`ngram` * `sequence_length`) which can be in the order of ~1000. Single tensor assignments, especially on accelerators, are inefficient.

This PR proposes a vectorized implementation which performs at most `ngram` assignments, which should be significantly lower than `ngram * sequence_length`.

A quick experiment shown at https://gist.github.com/guillaume-be/e6b862c701fac1b54765e7af7e71c641 shows that:
1. this `ngram_attention_bias` calculation is very expensive, taking close to 230ms (!) on a GPU
2. the vectorized implementation is several orders of magnitude faster (the same calculation takes less than 1ms on the same example)

## Who can review?

@patrickvonplaten maybe you would be a good candidate? I could not find anyone assigned for ProphetNet

edit: pushed some further optimization, further accelerating by ~40%

### CODE REVIEW

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 1]
```diff
@@ -171,13 +171,16 @@ def ngram_attention_bias(sequence_length, ngram, device, dtype):
...
+    right_block = left_block.detach().clone()
     # create bias
     for stream_idx in range(ngram):
-        for i in range(sequence_length):
-            bias[stream_idx, i, sequence_length + i] = 0
-            bias[stream_idx, i, : max(i - stream_idx, 0) + 1] = 0
-    return bias
+        right_block[stream_idx].fill_diagonal_(0, wrap=False)
+        left_block[stream_idx].triu_(-stream_idx + 1)
+        for i in range(stream_idx):
```
Comment (patrickvonplaten) [inline src/transformers/models/prophetnet/modeling_prophetnet.py:180]: could we maybe get rid of this loop as well?
Something like 

`left_block[stream_idx, torch.arange(stream_idx), 0]` should work I think

**Review by patrickvonplaten — COMMENTED** [review]
Thanks a lot @guillaume-be ! Very nice to remove the `for i in range(sequence_length)` call -> this does indeed look very suboptimal. 

I'm fine with the PR! When all changes are made I'll run the slow tests of ProphetNet locally to make sure nothing broke unexpectantly and merge then :-)

File: src/transformers/models/prophetnet/modeling_prophetnet.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/models/prophetnet/modeling_prophetnet.py]: Your new change is even better - nice :-)

Comment (guillaume-be) [reply src/transformers/models/prophetnet/modeling_prophetnet.py]: Yes, just noticed it as well. Actually pushing an even simpler version, speeding up further the calculation down to ~600us on my machine (400x faster than the baseline)

**Comment by patrickvonplaten** [comment]
All slow tests are passing! Very nice PR - thanks a mille @guillaume-be
